packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,290 @@
+import { uuid } from './uuid'
+import {
+  FetchHTTPClient,
+  HTTPClient,
+  HTTPClientRequest,
+  HTTPResponse,
+} from './http-client'
+import { SignOptions, sign } from 'jsonwebtoken'
+import { Emitter, backoff, sleep } from '@segment/analytics-core'
+import { AbortSignal, AbortController } from './abort'
+
+type AccessToken = {
+  access_token: string
+  expires_in: number
+}
+
+export type OAuthSettings = {
+  clientId: string
+  clientKey: Buffer
+  keyId: string
+  authServer?: string
+  scope?: string
+  maxRetries?: number
+  httpClient?: HTTPClient
+}
+
+export class TokenManager {
+  private alg = 'RS256' as const
+  private grantType = 'client_credentials' as const
+  private clientAssertionType =
+    'urn:ietf:params:oauth:client-assertion-type:jwt-bearer' as const
+  private clientId: string
+  private clientKey: Buffer
+  private keyId: string
+  private scope: string
+  private authServer: string
+  private httpClient: HTTPClient
+  private maxRetries: number
+  private clockSkewInSeconds = 0
+
+  private controller: AbortController
+  private signal: AbortSignal
+
+  private accessToken?: AccessToken
+  private tokenEmitter = new Emitter<{
+    access_token: [{ token: AccessToken } | { error: unknown }]
+  }>()
+  private retryCount: number
+  private lastError: any
+  private pollerTimer?: ReturnType<typeof setTimeout>
+
+  constructor(props: OAuthSettings) {
+    this.keyId = props.keyId
+    this.clientId = props.clientId
+    this.clientKey = props.clientKey
+    this.authServer = props.authServer ?? 'https://oauth2.segment.io'
+    this.scope = props.scope ?? 'tracking_api:write'
+    this.httpClient = props.httpClient ?? new FetchHTTPClient()
+    this.maxRetries = props.maxRetries ?? 3
+    this.tokenEmitter.on('access_token', (event) => {
+      if ('token' in event) {
+        this.accessToken = event.token
+      }
+    })
+    this.controller = new AbortController()
+    this.signal = this.controller.signal
+    this.signal.addEventListener('abort', () => {
+      clearTimeout(this.pollerTimer)
+    })
```

Comment:
It looks like the `AbortController` isn't needed anymore since it is only called by `stopPoller()`, and stopPoller is already calling `clearTimeout`